src/golang/cmd/server/handler/connect_integration.go

```diff
@@ -159,27 +163,53 @@ func (h *ConnectIntegrationHandler) Perform(ctx context.Context, interfaceArgs i
 		return emptyResp, statusCode, err
 	}
 
-	if args.SetAsStorage {
-		// This integration should be used as the new storage layer
-		if err := setIntegrationAsStorage(
-			ctx,
-			args.Service,
-			args.Config,
-			h.WorkflowDagReader,
-			h.WorkflowDagWriter,
-			h.ArtifactReader,
-			h.ArtifactResultReader,
-			h.OperatorReader,
-			h.Database,
-		); err != nil {
-			return emptyResp, http.StatusInternalServerError, errors.Wrap(err, "Unable to change metadata store.")
-		}
-	}
-
 	if err := txn.Commit(ctx); err != nil {
 		return emptyResp, http.StatusInternalServerError, errors.Wrap(err, "Unable to connect integration.")
 	}
 
+	if args.SetAsStorage {
+		// This integration should be used as the new storage layer.
+		// In order to do so, we need to migrate all content from the old store
+		// to the new store. This requires pausing the server and then restarting it.
+		// All of this logic is performed asynchronously so that the user knows that
+		// the connect integration request has succeeded and that the migration is now
+		// under way.
+		go func() {
+			// Wait until the server is paused
+			h.PauseServer()
+			// Makes sure that the server is restarted
+			defer h.RestartServer()
+
+			// Wait until there are no more workflow runs in progress
+			lock := utils.NewExecutionLock()
```

Comment:
We need to move this lock acquisition to this higher level function, because workflow runs should continue to remain blocked until the actual storage config has been modified. This happens at the end of the `setIntegrationAsStorage` function.